Tutorial implement a custom feature (#131)
ignitionrobotics/ign-physics: tutorials/08-implementing-a-custom-feature.md

    Every Engine feature **must** inherit this class.
  - \ref ignition::physics::Feature::Joint "Joint": defines physics concept `Joint` behaviors (for example
    the \ref ignition::physics::GetBasicJointState "GetBasicJointState"
    feature).
  - \ref ignition::physics::Feature::Link "Link": defines physics concept `Link` structure.
  - \ref ignition::physics::Feature::Model "Model": defines physics concept `Model` structure (for example
    the \ref ignition::physics::GetLinkFromModel "GetLinkFromModel"
    feature including both `Link` and `Model` objects).
  - \ref ignition::physics::Feature::Shape "Shape": defines physics concept `Shape` structure (for example
    the \ref ignition::physics::GetShapeKinematicProperties "GetShapeKinematicProperties"
    feature).
  - \ref ignition::physics::Feature::World "World": defines physics concept `Shape` structure (for example
    the \ref ignition::physics::dartsim::RetrieveWorld "RetrieveWorld" feature
    in `dartsim` plugin).

  Note that these object classes are not mutually exclusive and could be defined in conjunction together to describe the `Feature`. There are also other uncommon

>   objects defined depending on feature functionality, for example, the \ref ignition::physics::SetFreeGroupWorldPose::FreeGroup "FreeGroup"

Reviewer: these two lines can be split into shorter segments per line

  physics concepts, please refer to
  \ref physicsconcepts "Ignition Physics simulation concepts" tutorial.
- \ref ignition::physics::Feature::Implementation "Implementation" interfaces
the actual physics engines API for the custom feature. It has
\ref ignition::physics::Feature::Implementation::InitiateEngine "InitiateEngine"
 to trigger physics engine initiation to provide the required functionalities.

Moreover, we can define dependencies between custom `Features`:
- By default, a blank feature will not require any other features.
If the custom feature does require some other set of features,
then it should be inherited from
\ref ignition::physics::FeatureWithRequirements "FeatureWithRequirements" class,
and provided a list of the `Features` required.
- By default, a blank feature will not conflict with any other features. If
the custom feature does conflict with some other set of features, then it should
be inherited from
\ref ignition::physics::FeatureWithConflicts "FeatureWithConflicts" class,
and provided a list of the conflicting `Features`. The conflicting `Features`
will not run at the same time when requested.

### Define the custom feature

> 

Reviewer: It'd be helpful to establish a folder structure for user to follow along. So that they know what code block to add to which file throughout . Then in the end, they will have a working example folder which they can compile locally.

Reviewer: @anindex friendly ping on this one :)

Author: I did add a folder structure a few line below. Im not sure if that what you mean :)

Author: Okay I added the folder structure for this section :)

, we show it here for convenience:

```cpp
#include <dart/simulation/World.hpp>
#include <ignition/physics/FeatureList.hh>

namespace ignition {
namespace physics {
namespace dartsim {

/////////////////////////////////////////////////
class RetrieveWorld : public virtual Feature
{
  public: template <typename PolicyT, typename FeaturesT>
  class World : public virtual Feature::World<PolicyT, FeaturesT>
  {
    /// \brief Get the underlying dartsim world for this World object.
    public: dart::simulation::WorldPtr GetDartsimWorld();
  };

  public: template <typename PolicyT>
  class Implementation : public virtual Feature::Implementation<PolicyT>
  {
    public: virtual dart::simulation::WorldPtr GetDartsimWorld(
        const Identity &_worldID) = 0;
  };
};

/////////////////////////////////////////////////
template <typename PolicyT, typename FeaturesT>
dart::simulation::WorldPtr RetrieveWorld::World<PolicyT, FeaturesT>
::GetDartsimWorld()
{
  return this->template Interface<RetrieveWorld>()
      ->GetDartsimWorld(this->identity);
}

}
}
}
```

> As can be seen, after including the necessary library of `dartsim` and `ign-physics`,

Reviewer: ```suggestion
As seen above, after including the necessary library of `dartsim` and `ign-physics`,
```

};

}
}
}

```

The custom feature `RetrieveWorld` is added to `CustomFeatureList`, other custom
features could also be added here.
The `CustomFeatures` "FeatureList" here inherits from:
- [Base](https://github.com/ignitionrobotics/ign-physics/blob/main/dartsim/src/Base.hh)
class for foundation definitions of Models, Joints, Links, and Shapes objects
of `dartsim` interfacing to Ignition Physics API.
- \ref ignition::physics::Implements3d "Implements3d" for implementing the
custom feature with \ref ignition::physics::FeaturePolicy3d "FeaturePolicy3d"
("FeaturePolicy" of 3 dimensions and scalar type `double`).

Based on the `CustomFeatureList`, we will override the corresponding member functions
declared in each of the custom features. In this case, we have only the custom
feature `RetrieveWorld` and its corresponding `Implementation::GetDartsimWorld`
member function.

- `CustomFeatures.cc`:

```cpp
#include "CustomFeatures.hh"

namespace ignition {
namespace physics {
namespace dartsim {

/////////////////////////////////////////////////
dart::simulation::WorldPtr CustomFeatures::GetDartsimWorld(
    const Identity &_worldID)
{
  return this->worlds.at(_worldID);
}

}
}
}
```

> 

Reviewer: At this point, it'd be helpful to show what files are included in this particular plugin folder, just as a sanity check for users to follow along, before they proceed to compile and test the new feature.
